Add retry limiter support

friscoMad opened the pull request:
Motivation:

As we discussed over discord there is currently no support for add proper retry limiting, this PR adds it with a couple of working retry limiter implementations, one based on GRPC logic and another that uses a RateLimiter.

Modifications:

- Added some methods to be able to set a RetryLimiter while building a RetryConfig
- Hooked the retry limiter inside RetryClient
- Created a RetryLimiter interface and a couple of implementations.

Result:

- Closes #6282 
- There are changes in public APIs in RetryConfig constructor and protected getNextDelay methods in AbstractRetryingClient
- It will allow clients to add a retry throttling logic if needed.

friscoMad committed:
Initial version of the RetryRateLimiter API and implementation

friscoMad committed:
Added executor to handle exceptions
Added tests
Added JavaDoc

friscoMad committed:
Fix JavaDoc

friscoMad committed:
Revert unintended changes

schiemon commented on core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java:
```diff
@@ -330,6 +332,8 @@ private void doExecute0(ClientRequestContext ctx, HttpRequestDuplicator rootReqD
 
         final RetryConfig<HttpResponse> config = mappedRetryConfig(ctx);
         if (!ctx.exchangeType().isResponseStreaming() || config.requiresResponseTrailers()) {
+            RetryLimiterExecutor.onCompletedAttempt(config.retryLimiter(), ctx, ctx.log().partial(),
```
I agree with the convoluted part; I am about to clean the retrying clients here: #6292 (on halt atm because I have exams)
So I think we can let that PR tackle this

friscoMad commented on core/src/main/java/com/linecorp/armeria/client/retry/limiter/GrpcRetryLimiter.java:
```diff
@@ -0,0 +1,205 @@
+     * @throws IllegalArgumentException if any argument is invalid
+     */
+    public GrpcRetryLimiter(int maxTokens, int tokenRatio, int threshold,
+                            Collection<Integer> retryableStatuses) {
```
This lives in the core package that does not depend on gRPC, and I think we want it that way. I could move the implementation to the grpc package, but then we can not have a `RetryLimiter.ofGrpc`, to me it makes more sense, but I tried to follow the pattern of the gRPC retry rule.

schiemon replied:
@friscoMad Maybe we can move `GrpcRetryLimiter` to the gRPC package and offer a static `GrpcRetryLimiter.of` builder?

friscoMad replied:
We can do that, but I would like to hear from @ikhoon as he was the one asking for the builders on RetryLimiter class.
For example grpc retry rule is in core instead of the grpc package, which makes more sense, so I was not sure what to do.

friscoMad committed:
Address comments

schiemon commented on core/src/main/java/com/linecorp/armeria/client/retry/AbstractRetryingClient.java:
```diff
@@ -214,7 +217,8 @@ protected final long getNextDelay(ClientRequestContext ctx, Backoff backoff) {
      */
     @SuppressWarnings("MethodMayBeStatic") // Intentionally left non-static for better user experience.
-    protected final long getNextDelay(ClientRequestContext ctx, Backoff backoff, long millisAfterFromServer) {
+    protected final long getNextDelay(ClientRequestContext ctx, @Nullable RetryLimiter limiter,
```
As this class is also not final people could inherit from it and use those methods

friscoMad committed:
Fix conditions and expected errors

friscoMad commented on core/src/main/java/com/linecorp/armeria/client/retry/limiter/GrpcRetryLimiter.java:
```diff
@@ -0,0 +1,208 @@
+ *
+ * <p>The implementation is heavily based on GRPC Java <a href="https://github.com/grpc/grpc-java/blob/94532a6b56076c56fb9278e9195bba1190a9260d/core/src/main/java/io/grpc/internal/RetriableStream.java#L1463">implementation</a>.
+ */
+public class GrpcRetryLimiter implements RetryLimiter {
```
This implementation tries to mimic gRPC Java as much as possible. There are 2 points that are not implemented. 
gRPC java has support for [pushback](https://github.com/grpc/proposal/blob/4f833c5774e71e94534f72b94ee1b9763ec58516/A6-client-retries.md?plain=1#L201) header, as that is not implemented in the RetryingClient, I have not added support for it here.

The other difference is that gRPC Java considers errors (for reducing tokens) every status defined in the hedging policy as [nonFatalStatusCodes](https://github.com/grpc/proposal/blob/4f833c5774e71e94534f72b94ee1b9763ec58516/A6-client-retries.md?plain=1#L193). We don't have hedging implemented, but the same behaviour can be replicated with the extended constructor passing the complete list of statuses.

schiemon commented on core/src/main/java/com/linecorp/armeria/client/retry/limiter/GrpcRetryLimiter.java:
```diff
@@ -0,0 +1,205 @@
+    final int threshold;
+
+    /**
+     * 1000 times the tokenRatio field of the retryThrottling policy in service config.
```
Yeah I misunderstood the scaling part (see below)

friscoMad commented on core/src/main/java/com/linecorp/armeria/client/retry/limiter/GrpcRetryLimiter.java:
```diff
@@ -0,0 +1,208 @@
+     * @param maxTokens Initial token count
+     * @param tokenRatio Number of tokens a successful request increments
+     */
+    public GrpcRetryLimiter(int maxTokens, int tokenRatio) {
```
I missed the float on the constructors, ok, I will need to split the constructors and update some tests. I will do that tomorrow. Thanks for the heads up, I totally missed that part.

friscoMad replied:
Ok, it should be fixed now!

friscoMad committed:
Change constructor API to support floats
Refactor constructors
Update javadoc

schiemon commented on core/src/main/java/com/linecorp/armeria/client/retry/limiter/RetryLimiter.java:
```diff
@@ -0,0 +1,90 @@
+     * @param maxTokens the initial token count (must be positive)
+     * @param tokenRatio the number of tokens a successful request increments (must be positive)
+     */
+    static RetryLimiter ofGrpc(int maxTokens, int tokenRatio) {
```
:)

friscoMad replied:
🤦  Why is this not breaking the build? It should fail to compile

schiemon replied:
It is actually legal code:
https://onecompiler.com/jshell/43rjpddqw

schiemon reviewed (commented):
Did another pass and only have a few nits

schiemon commented on core/src/test/java/com/linecorp/armeria/client/retry/limiter/RetryRateLimiterTest.java:
```diff
@@ -0,0 +1,179 @@
+        // Second retry immediately after should be denied due to rate limiting
+        assertThat(limiter.shouldRetry(ctx, 3)).isFalse();
+
+        // Wait for 2 seconds to allow rate limiter to refill (0.5 permits per second = 2 seconds per permit)
+        try {
+            Thread.sleep(2100);
```
```suggestion
        // Wait for 2 seconds + some tolerance to allow rate limiter to refill (0.5 permits per second = 2 seconds per permit)
        try {
            Thread.sleep(2000 + 100);
```

schiemon commented on core/src/main/java/com/linecorp/armeria/client/retry/limiter/RetryLimiterExecutor.java:
```diff
@@ -0,0 +1,101 @@
+ * and doing nothing for {@link #onCompletedAttempt}).
+ * </p>
+ */
+public final class RetryLimiterExecutor {
```
I would not expose this kind of utility class

friscoMad replied:
Fixed

schiemon commented on core/src/test/java/com/linecorp/armeria/client/retry/limiter/RetryRateLimiterTest.java:
```diff
@@ -0,0 +1,179 @@
+            executor.submit(() -> {
+                try {
+                    for (int j = 0; j < attemptsPerThread; j++) {
+                        if (limiter.shouldRetry(ctx, j + 2)) {
```
Why +2?

friscoMad committed:
Moved RetryLimiterExecutor to change it to package private
Fixed RetryLimiter.of constructors
Added test for RetryLimited.of constructors
Improved logs
Added toString methods
Checked Rate limiter param

friscoMad committed:
Improve docs
Add toString docs

schiemon reviewed (commented):
Sorry for the incremental reviews but found a few important points here😆

friscoMad committed:
Address fixes

friscoMad committed:
Fix style

friscoMad commented on core/src/main/java/com/linecorp/armeria/client/retry/limiter/GrpcRetryLimiter.java:
```diff
@@ -0,0 +1,238 @@
+        }
+        // Extract the headers to be able to evaluate the gRPC status
+        // Check HTTP trailers first, because most gRPC responses have non-empty payload + trailers.
+        HttpHeaders maybeGrpcTrailers = requestLog.responseTrailers();
```
Changed 👍

jrhee17 committed:
Merge branch 'main' into ramiro-aparicio/Add-RetryLimiter

jrhee17 reviewed (commented):
Sorry about the late review, some rough thoughts I had while taking an initial look:

- `RetryConfig#requiresResponseTrailers` dictates whether to wait for trailers before deciding whether to retry or not. Given the current implementation, I'm unsure if `GrpcRetryLimiter` will be able to consistently retrieve grpc trailers for streaming responses. (e.g. If `RetryRule.RetryRule.onUnprocessed()` is used, `RetryLimiter#onCompletedAttempt` may be invoked before the trailers arrive for a streaming response). In general (and not just gRPC), I guess users will have to keep track of what properties will be available at `RetryLimiter` based on how the `RetryConfig#requiresResponseTrailers` is set.

- When determining whether a retry should be done or not, if users would like to retry on a certain `grpc-status`, it seems like he/she should set both the `GrpcRetryLimiter` and the `RetryRule` since they seem to have an AND relationship. (if they would like to use `GrpcRetryLimiter`). In general, I'm still trying to organize the exact distinction `RetryRule` and `RetryLimiter.`

e.g.
```
RetryRule.builder()
                 .onGrpcTrailers((ctx, trailers) -> trailers.containsInt("grpc-status", 15))
...
new GrpcRetryLimiter(0, 0, 0, List.of(15))
```

- Maybe nit, but it seems like `RetryLimiter#shouldRetry` may be called even if a retry won't actually execute (e.g. due to `setResponseTimeout` checks, request abortion, etc.). While it may not be a big deal, this may be a little awkward for the `RetryRateLimiter` implementation which acquires.

I'm nitpicking because I'm still trying to grasp the full picture at the moment.
I would also like to do some research on gRPC's implementation (as well as envoy's) to have a better idea on retry budgets in general before leaving more comments.

friscoMad commented:
> Sorry about the late review, some rough thoughts I had while taking an initial look:
> 
>     * `RetryConfig#requiresResponseTrailers` dictates whether to wait for trailers before deciding whether to retry or not. Given the current implementation, I'm unsure if `GrpcRetryLimiter` will be able to consistently retrieve grpc trailers for streaming responses. (e.g. If `RetryRule.RetryRule.onUnprocessed()` is used, `RetryLimiter#onCompletedAttempt` may be invoked before the trailers arrive for a streaming response). In general (and not just gRPC), I guess users will have to keep track of what properties will be available at `RetryLimiter` based on how the `RetryConfig#requiresResponseTrailers` is set.

Yes, and this is difficult even for me trying to understand how the code works, I am not sure if there is a reason for having a gRPC retry limiter without a gRPC retry rule, but maybe we can make some changes to make that decision explicit in RetryLimiter and enable `requiresResponseTrailers` if the limiter wants it.
> 
>     * When determining whether a retry should be done or not, if users would like to retry on a certain `grpc-status`, it seems like he/she should set both the `GrpcRetryLimiter` and the `RetryRule` since they seem to have an AND relationship. (if they would like to use `GrpcRetryLimiter`). In general, I'm still trying to organize the exact distinction `RetryRule` and `RetryLimiter.`
RetryRule decides if a retry shoud be done, retry limiter can block a retry based on whatever rules are implemented (like max retries in flight, max retries per minute, token bucket...) the retry limiter only is triggered if a retry is going to be done.
GrpcRetryLimiter tries to mimic the Grpc retry throttling logic defined here: https://github.com/grpc/proposal/blob/master/A6-client-retries.md#throttling-retry-attempts-and-hedged-rpcs it is expected to be used with a GrpcRetry rule, but the rule only covers the retriable statuses of the [grpc retry policy](https://github.com/grpc/proposal/blob/master/A6-client-retries.md#retry-policy-capabilities). Everything else is included in parts of the armeria retry config.
> 
> 
> e.g.
> 
> ```
> RetryRule.builder()
>                  .onGrpcTrailers((ctx, trailers) -> trailers.containsInt("grpc-status", 15))
> ...
> new GrpcRetryLimiter(0, 0, 0, List.of(15))
> ```
> 
>     * Maybe nit, but it seems like `RetryLimiter#shouldRetry` may be called even if a retry won't actually execute (e.g. due to `setResponseTimeout` checks, request abortion, etc.). While it may not be a big deal, this may be a little awkward for the `RetryRateLimiter` implementation which acquires.
> 
This is a valid concern and I am happy to move the check to another point in the timeline, I have a lot of difficulties to fully understand the flow and find where to add the 2 hooks (`shouldRetry` and `onCompletedAttempt`), it is true that we want to check only when we are going to do the retry but also waiting for the backoff just to cancel the retry is a bit weird. I think that as long as abortions and errors are flows that are exceptional and not frequently used, I think it is fine to miscount by 1 from time to time. But I have not an strong opinion we can move the check to be done as late as possible and that will be fine for me too.
> 
> I'm nitpicking because I'm still trying to grasp the full picture at the moment. I would also like to do some research on gRPC's implementation (as well as envoy's) to have a better idea on retry budgets in general before leaving more comments.
No worries, it is a very niche topic, and my feeling is that with all the edge cases that can happen with IO, delays... we can only expect to do it "good enough".

jrhee17 reviewed (commented):
I'm still organizing my thoughts, but I think overall looks good. Left some minor comments (mostly to show that I'm not forgetting about this PR)

jrhee17 commented on core/src/main/java/com/linecorp/armeria/client/retry/limiter/GrpcRetryLimiter.java:
```diff
@@ -0,0 +1,238 @@
+     */
+    final int tokenRatio;
+
+    final Set<String> retryableStatuses;
```
The idea wasn't to implement throttling based on `RetryRule` - `RetryLimiter#canRetry` can still determine whether a retry should be throttled or not.

The idea was rather to add a property to RetryDecision which allows users to specify whether a `retryDecision` should affect throttling
```java
final Set<String> retryableStatusCodes = ImmutableSet.of("14");
RetryRule rule = new RetryRule() {
    @Override
    public CompletionStage<RetryDecision> shouldRetry(ClientRequestContext ctx,
                                                      @Nullable Throwable cause) {
        final RequestLog headers = ctx.log().getIfAvailable(RequestLogProperty.RESPONSE_HEADERS);
        if (cause instanceof UnprocessedRequestException || headers == null) {
            return RetryDecision.of(backoff, NO_THROTTLE);
        }
        final String status = headers.responseHeaders().get("grpc-status");
        if (status == null) {
            return RetryDecision.noRetry(THROTTLE_SUCCESS);
        } else if (!retryableStatusCodes.contains(status)) {
            return RetryDecision.of(backoff, NO_THROTTLE);
        } else {
            return RetryDecision.of(backoff, THROTTLE_FAILURE);
        }
    }
};
```

and users can still decide on how to limit retries based on the throttling decision

```java
interface RetryLimiter {
  boolean shouldRetry(ClientRequestContext ctx, int numAttemptsSoFar);
  default void onCompletedAttempt(RetryDecision decision, ClientRequestContext ctx, RequestLog requestLog, int numAttemptsSoFar);
}
```

friscoMad replied:
Ohh Ok, I understand your idea, but it seems more complex and require more coupling.
Afaik, usually both subsystems can be configured independently, and while in general solutions only provide a single algorithm for retry and retry budget/throttling, so there is more space for coupling, but in Armeria it seems that you are open to creating custom implementations, so I think decoupling is better, that said I understand the benefits of doing something like that for the gRPC rule/limitter as the official implementation is more complex than other limiters, maybe we can fix that (checking the status twice, setting retriable statuses in both cases) by using some context magic?

jrhee17 replied:
> maybe we can fix that (checking the status twice, setting retriable statuses in both cases) by using some context magic?

I'm not sure if there is an easy way to do this with the current set of `RetryRule` APIs, but if users can specify `retryableStatusCodes ` once and have it applied to both the retry rule and limiter, I have no issue with the current API proposal.

In general (not just gRPC), I think it's possible that users will want to apply throttling for only a subset of responses
(e.g. 429 too many requests, 503 gateway error, etc..) so I think if there is a general solution for this, it would be great as well.

Also, I'm not sure if we'll want the flexibility for users to also throttle based on response content (e.g. imagining cases where server returns the exception in the response `{"error": "ConflictException"}`, `{"error": "ThrottledException"}`) which I think is slightly less common. I'm curious if the current proposed API can support this use-case (or if it should be supported at all)

ikhoon replied:
I slightly lean toward making `RetryLimiter` work independently to achieve decoupling. 

> ```java
> RetryRule customRetryRule = RetryRule.onUnprocessed()
>         .orElse(RetryRule.builder()
>                          .onGrpcTrailers((ctx, headers) -> > retriableStatuses.contains(headers.getInt("grpc-status")))
>                          .thenBackoff());
> RetryConfig.builder(customRetryRule)
>            .limiter(RetryLimiter.ofGrpc(1, 1, 1, retriableStatuses)).build();
> ```
> In general (not just gRPC), I think it's possible that users will want to apply throttling for only a subset of responses (e.g. 429 too many requests, 503 gateway error, etc..) so I think if there is a general solution for this, it would be great as well so RetryRateLimiter can also benefit from this.


In this case, the default implementation, which decreases the quota for each attempt, may cover it. 

```java
// The default `RetryLimiter` respects the retry decision made in customRetryRule
RetryConfig.builder(customRetryRule)
           .limiter(RetryLimiter.ofDefault(maxTokens, ...)).build();
```

jrhee17 replied:
I see, if you don't think this is inconvenient for users, feel free to proceed with the current implementation. Let me revisit this PR later once my review comments are addressed

jrhee17 committed:
Merge branch 'main' into ramiro-aparicio/Add-RetryLimiter
